### PR TITLE
[WIP] Adds Split-Explicit to HydrostaticFreeSurfaceModel 

### PR DESCRIPTION
**Pull request by sandreza**
This is the first step towards #2012. The goal is to test the method with a flat-bottom and compare performance / stability with respect to the other free-surface implementations.

### CODE REVIEW

**Comment by sandreza** [comment]
This is the tentative plan: 
1. Test the Split-Explicit time-stepping in standalone mode without the 3D model
2. Write kernels and tests for computing vertical integrals of tendency terms
3. Write the AB extrapolation function for the tendency terms
4. Write the "reconciliation" kernel that takes away the barotropic component of the velocity field and adds in the new one
5. Test everything together in various channel settings

File: src/Models/HydrostaticFreeSurfaceModels/split_explicit_free_surface.jl [hunk 1]
```diff
@@ -95,8 +95,8 @@ SplitExplicitAuxiliary{𝒞ℱ, ℱ𝒞}
 # Members
 `Gᵁ` : (ReducedField). Vertically integrated slow barotropic forcing function for U
 `Gⱽ` : (ReducedField). Vertically integrated slow barotropic forcing function for V
-`Hᶠᶜ`: (ReducedField). Depth at (Face, Center)
-`Hᶜᶠ`: (ReducedField). Depth at (Center, Face)
+`Hᶠᶜ`: (ReducedField). Depth at (Face, Center): minimum depth of neighbors
+`Hᶜᶠ`: (ReducedField). Depth at (Center, Face): minimum depth of neighbors
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/split_explicit_free_surface.jl:103]: The comment is slightly misleading; this is only the case for some types of immersed boundaries

**Comment by sandreza** [comment]
Just adding here as a small reminder: 

Currently the split-explicit scheme does not seem to conserve to machine precision the free-surface height average (as it should). In the current it only conserves to 8 digits. This is only when it is implemented as a part of the hydrostatic model. In a standalone test it does seem to conserve to machine precision. Once this issue is isolated and fixed I'd say it's good to go for merging

**Comment by christophernhill** [comment]
> Just adding here as a small reminder:
> 
> Currently the split-explicit scheme does not seem to conserve to machine precision the free-surface height average (as it should). In the current iteration it only conserves to 8 digits. This is only when it is implemented as a part of the hydrostatic model. In a standalone test it does seem to conserve to machine precision. Once this issue is isolated and fixed I'd say it's good to go for merging

@sandreza are you using AB time-stepping in the standalone test?

**Comment by sandreza** [comment]
> 

With the geostrophic adjustment test the large timesteps are computed with AB2. After doing some more testing it seems that the lack of conservation has been isolated to a topology with bounded directions. In the fully periodic case the eta field mean is conserved to machine precision

File: test/test_split_explicit_free_surface_solver.jl [hunk 2]
```diff
@@ -11,13 +11,14 @@ import Oceananigans.Models.HydrostaticFreeSurfaceModels: SplitExplicitFreeSurfac
 import Oceananigans.Models.HydrostaticFreeSurfaceModels: SplitExplicitState, SplitExplicitAuxiliary, SplitExplicitSettings, split_explicit_free_surface_substep!
```
Comment (glwagner) [inline test/test_split_explicit_free_surface_solver.jl:12]:
```suggestion
include("dependencies_for_runtests.jl")
```

File: test/test_split_explicit_free_surface_solver.jl [hunk 3]
```diff
@@ -11,13 +11,14 @@ import Oceananigans.Models.HydrostaticFreeSurfaceModels: SplitExplicitFreeSurfac
...
 
 @testset "Split-Explicit Dynamics" begin
-    arch = Oceananigans.CPU()
+
+for arch in [Oceananigans.GPU(), Oceananigans.CPU()]
```
Comment (glwagner) [inline test/test_split_explicit_free_surface_solver.jl:15]:
```suggestion
for arch in archs
```

**Comment by sandreza** [comment]
> * Test the Split-Explicit time-stepping in standalone mode without the 3D model
> * Write kernels and tests for computing vertical integrals of tendency terms
> * Write the AB extrapolation function for the tendency terms
> * Write the "reconciliation" kernel that takes away the barotropic component of the velocity field and adds in the new one
> * Test everything together in various channel settings

This has all been done

**Review by simone-silvestri — APPROVED** [review]

**Pull request merged**